### PR TITLE
[SimplifyCFG] Don't use a mask for lookup tables generated from switches with an unreachable default case

### PR DESCRIPTION
**Pull request by DaMatrix**
When transforming a switch with holes into a lookup table, we currently use a mask to check if the current index is handled by the switch or if it is a hole. If it is a hole, we skip loading from the lookup table. Normally, if the switch's default case is unreachable this has no impact, as the mask test gets optimized away by subsequent passes. However, if the switch is large enough that the number of lookup table entries exceeds the target's register width, we won't be able to fit all the cases into a mask and the switch won't get transformed into a lookup table. If we know that the switch's default case is unreachable, we know that the mask is unnecessary and can skip constructing it entirely, which allows us to transform the switch into a lookup table.

[Example](https://godbolt.org/z/7x7qfx8M1)

In the future, it might be interesting to consider allowing lookup table masks to be more than one register large (e.g. using a constant array of bit flags, similar to `std::bitset`).

### CODE REVIEW

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6750]:
```suggestion
  bool DefaultIsReachable = !SI->defaultDestUndefined();
```

Comment (DaMatrix) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp]: Changed! Didn't realize that was a thing, I just copied the existing code :P

**Comment by DianQK** [comment]
Please rebase your PR, see: https://github.com/llvm/llvm-project/commit/07b9d231ff9baa6473b0dd588a3ce5330d3e4871.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (DianQK) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6752]: This variable name is confusing to me. Could you explain it further and add comments?

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (DianQK) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6899]: When the default branch is undefined, we can use a poison value. This can be done as another PR, as it requires modifying the constructor of `SwitchLookupTable`.

Comment (DianQK) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6899]: But I'm not sure if this makes sense. :)

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll
Comment (DianQK) [inline llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll:5]: Is it possible to reduce the IR using 32 bits?

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (DaMatrix) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6756]: I've pushed some new comments explaining this in more detail.

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll
Comment (DaMatrix) [reply llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll]: Very good point, I didn't even think of that! I've done that in the latest revision.

**Comment by DaMatrix** [comment]
> Please rebase your PR, see: [07b9d23](https://github.com/llvm/llvm-project/commit/07b9d231ff9baa6473b0dd588a3ce5330d3e4871).

@DianQK I've rebased onto that, but it doesn't affect this PR since `SimplifyCFG/X86/switch_to_lookup_table.ll` doesn't appear to have been regenerated by that commit.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (DianQK) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6756]: Maybe you could change it to `AllHolesAreUndefined`, `AllHolesArePoison` or `AllHolesAreUnreachable`.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (DianQK) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6910]: I believe this makes sense. For example, we could treat `[1, poison, 3, 4]` as `[1, 2, 3, 4]` or merge `[1, poison, 3, 4]` and `[1, 2, poison, 4]` into `[1, 2, 3, 4]`. If you're considering implementing it, please let me know. :)

**Review by dianqk — APPROVED** [review]
LGTM. If no one else comments, I'll merge it after next Monday.

**Comment by DianQK** [comment]
> > Please rebase your PR, see: [07b9d23](https://github.com/llvm/llvm-project/commit/07b9d231ff9baa6473b0dd588a3ce5330d3e4871).
> 
> @DianQK I've rebased onto that, but it doesn't affect this PR since `SimplifyCFG/X86/switch_to_lookup_table.ll` doesn't appear to have been regenerated by that commit.

Sorry, I didn't notice. You can manually discard these changes.

**Comment by nikic** [comment]
> > > Please rebase your PR, see: [07b9d23](https://github.com/llvm/llvm-project/commit/07b9d231ff9baa6473b0dd588a3ce5330d3e4871).
> > 
> > 
> > @DianQK I've rebased onto that, but it doesn't affect this PR since `SimplifyCFG/X86/switch_to_lookup_table.ll` doesn't appear to have been regenerated by that commit.
> 
> Sorry, I didn't notice. You can manually discard these changes.

I've regenerated the file now.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6756]:
```suggestion
  bool HolesHaveUndefinedResults = TableHasHoles && !HasDefaultResults;
```

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:6764]:
```suggestion
  bool NeedMask = HolesHaveUndefinedResults && DefaultIsReachable;
```

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll
Comment (nikic) [inline llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll:2]: I don't think `-keep-loops` is relevant here? We also don't need both RUN lines, they do the same thing.

**Review by nikic — APPROVED** [review]
LGTM as well.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp
Comment (DaMatrix) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp]: My bad, I was keeping with the style of the existing code. I've removed these parens (and the other ones you pointed out)

File: llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll
Comment (DaMatrix) [reply llvm/test/Transforms/SimplifyCFG/X86/switch_to_lookup_table_big.ll]: Alright, I've removed `-keep-loops=false` and the second `RUN`. I copied those lines from the original `switch_to_lookup_table.ll`; I assumed there was some reason why both versions were being tested.

**Comment by DaMatrix** [comment]
https://github.com/llvm/llvm-project/pull/94468#discussion_r1630381209 I like `AllHolesAreUndefined`, have changed it to that.

https://github.com/llvm/llvm-project/pull/94468#discussion_r1630383915 Yeah, I'm definitely interested in doing that, I'll get working on that as soon as this is merged. @DianQK

**Comment by DianQK** [comment]
> [#94468 (comment)](https://github.com/llvm/llvm-project/pull/94468#discussion_r1630381209) I like `AllHolesAreUndefined`, have changed it to that.
> 
> [#94468 (comment)](https://github.com/llvm/llvm-project/pull/94468#discussion_r1630383915) Yeah, I'm definitely interested in doing that, I'll get working on that as soon as this is merged. @DianQK

Thanks. I prefer `AllHolesArePoison` because unreachable immediately results in undefined behavior, so we can use `poison`, which is stronger than `undef`. We can discuss it in the next PR. :)

**Pull request merged**
